pfuture.el

```diff
@@ -175,35 +225,58 @@ details see documentation of `accept-process-output'."
      process timeout nil just-this-one))
   (process-get process 'result))
 
+(cl-macrolet
+    ((define-getter (name doc variable )
+       `(define-inline ,name (process)
+          ,doc
+          (declare (side-effect-free t))
+          (inline-letevals (process)
+            (inline-quote
+             (process-get ,',process ',variable))))))
+  (define-getter pfuture-result "Return the output of a pfuture PROCESS." stdout)
+  (define-getter pfuture-stderr "Return the error output of a pfuture PROCESS." stderr))
+
 (defun pfuture-await-to-finish (process)
   "Keep reading the output of PROCESS until it is done.
 Same as `pfuture-await', but will keep reading (and blocking) so long as the
 process is *alive*.
 
 If the process never quits this method will block forever. Use with caution!"
+  ;; If the sentinel hasn't run, disable it. We are going to delete
+  ;; the stderr process here.
+  (set-process-sentinel process nil)
   (let (inhibit-quit)
-    (accept-process-output process nil nil t)
-    (while (process-live-p process)
-      (accept-process-output process nil nil t)))
-  (process-get process 'result))
-
-(define-inline pfuture-result (process)
-  "Return the output of a pfuture PROCESS."
-  (declare (side-effect-free t))
-  (inline-letevals (process)
-    (inline-quote
-     (process-get ,process 'result))))
-
-(defun pfuture--append-output (process msg)
-  "Append PROCESS' MSG to the already saved output."
-  (process-put process 'result (concat (process-get process 'result) msg)))
+    (while (accept-process-output process)))
+  (let* ((plist (process-plist process))
+         (stderr-process (plist-get plist 'stderr-process)))
+    (when stderr-process
+      (pfuture--delete-process stderr-process))
+    (plist-get plist 'stdout)))
 
 (defun pfuture--append-output-to-buffer (process msg)
   "Append PROCESS' MSG to its output buffer."
   (with-current-buffer (process-get process 'buffer)
     (goto-char (point-max))
     (insert msg)))
 
+(defun pfuture--append-stdout (process msg)
+  "Append PROCESS' MSG to the already saved stdout output."
+  (let* ((process-plist (process-plist process))
+         (previous-output (plist-get process-plist 'stdout)))
+    (plist-put process-plist 'stdout
+               (if (zerop (string-bytes previous-output))
```

Comment:
In the usual case, this gets called only one time, so this optimization seems to be worth it.